Initialize dataAddr field only for non-zero length array

VermaSh opened the pull request:
Update array inline allocation sequence to initialize dataAddr field only for non-zero length arrays. Field should be left blank for zero length arrays. Additionally this also clears padding field after the size field.

VermaSh commented:
[Off-heap personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/25627/), [upstream personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/25629/). My off-heap personal build is looking good so far, only 1 unrelated failure. I'll remove the WIP tag once both builds have passed without off-heap related failures.

cc: @r30shah / @zl-wang

VermaSh commented:
Launched new [non-off-heap personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/25650/) to verify PR changes

VermaSh commented:
Relaunched non off-heap [build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/25712/)

VermaSh commented:
I have verified the sequence for x and z through a unit test. Need to do the same for power and aarch64 before removing WIP tag.

VermaSh commented:
Launched personal non off-heap [build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/25767/) for power.

VermaSh commented:
Kept only z here as it is ready for review, other platforms require more testing. @r30shah Can I please get a review?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -4948,33 +4948,28 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+      if (isOffHeapAllocationEnabled)
+         {
+         TR_ASSERT_FATAL_WITH_NODE(node,
+            TR::InstOpCode::LGF == loadDimLenOpCode,
```
Can you explain why we have this assert (Checking loadDimLenOpCode to `LGF`). That variable is initialized to `LGF` at line 4891 and I do not see it ever updated.

VermaSh replied:
`dataAddr` field should be 0 for 0 size arrays. Since size will already be 0, I figured we can use that to 0-out `dataAddr` field. The issue is that `size` is 4 bytes wide whereas `dataAddr` is 8 bytes wide, in order to use the same register I need to clear top 4 bytes of `firstDimLenReg`. I added the assert to ensure that size is sign extended to 64 bits when being loaded to `firstDimLenReg`. Same logic for second `LGF` assert.

r30shah replied:
I do get the reason behind using the LGF instruction. Do you mean to say the logic for the assert is safeguarding someone accidentally changing the code to use other instruction to load  firstDimLenReg ? If so instead of putting this assert which I feel is not needed, we should put the comment where you are loading firstDimLenReg on why using LGF instead of LG and warn about the potential functional issue.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5056,26 +5078,29 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
-      }
+      if (isOffHeapAllocationEnabled)
+         {
+         TR_ASSERT_FATAL_WITH_NODE(node,
```
Same question as before.

r30shah reviewed (changes requested):
Checking out the changes, by skimming through the change, I had some intiail questions, I have posted.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -4948,33 +4952,23 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+   if (TR::Compiler->om.compressObjectReferences())
+      {
+      // Clear padding in contiguous array header layout. +4 because size field is 4 bytes wide.
+      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, static_cast<int32_t>(fej9->getOffsetOfDiscontiguousArraySizeField()) + 4, cg));
```
Comment says clear padding but you are storing the `firstDimLenReg` for the Array Size. Can you clarify and eventually clean-up comment if needed.

Also why is the inconsistencies with static_cast throughout the code. May be remove it from here (There is anyway implicit cast to int32_t).

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -4948,33 +4952,23 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+      if (isOffHeapAllocationEnabled)
+         {
+         cursor = generateRXInstruction(cg, TR::InstOpCode::STG, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousDataAddrField(), cg));
+         iComment("Clear 1st dim dataAddr field.");
```
Same comment as before.

VermaSh replied:
Because we are dealing with 0 size array and we used `LGF` to load dimension size, `firstDimLenReg` will be 0 so I figured we could use that to clear `dataAddr` field.

r30shah replied:
Ok. This does feel like an implicit assumption. If the register gets modified in future, we end up with wrong code and this is tied with the use of `LGF` to use the dimension register. I am ok with this change. But make sure that we write this explicit assumption (That dimension register should contain NULL)

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5026,6 +5020,33 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
    cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, firstDimLenReg, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg));
    iComment("Init 1st dim size field.");
+   if (!TR::Compiler->om.compressObjectReferences())
+      { // Clear padding in contiguous array header layout. +4 because size field is 4 bytes wide.
```
Add this comment to new line, not besides braces.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5056,26 +5077,24 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+   if (TR::Compiler->om.compressObjectReferences())
+      {
+      // Clear padding in contiguous array header layout. +4 because size field is 4 bytes wide.
+      cursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, secondDimLenReg, generateS390MemoryReference(temp2Reg, static_cast<int32_t>(fej9->getOffsetOfDiscontiguousArraySizeField()) + 4, cg));
```
Same question as before with firstDimLenReg.

VermaSh replied:
Because we are dealing with 0 size array at this point, and we used `LGF` to load dimension size, `secondDimLenReg` will be 0 so I figured we could use that to clear `dataAddr` field.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10762,10 +10748,14 @@ genInitArrayHeader(TR::Node * node, TR::Instruction *& iCursor, bool isVariableL
       iCursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, eNumReg,
                       generateS390MemoryReference(resReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), iCursor);
+      // Clear padding in contiguous array header layout. +4 because size field is 4 bytes wide.
+      iCursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, eNumReg, generateS390MemoryReference(resReg, static_cast<int32_t>(fej9->getOffsetOfDiscontiguousArraySizeField()) + 4, cg), iCursor);
```
Can you explain the comment ?

VermaSh replied:
Extra padding is added to the array header for 8 byte boundary alignment; so with that comment I just meant that we add 4 to size field offset to get to the padding. I'll update the comment to be more clear.

r30shah replied:
Okk, does `enumReg`  contains 0 ?

VermaSh replied:
So this works because if `enumReg` isn't 0 (i.e. we aren't dealing with 0 size array), whatever we write at `getOffsetOfDiscontiguousArraySizeField` will be overwritten by the element or dataAddr field (if present). But I think using `XC` might provide better readability. I'll update the code.

VermaSh commented:
@r30shah Can I please get another review?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11130,69 +11120,80 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+               // Load address of first array element
+               iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, dataSizeReg, dataAddrMR, iCursor);
+               // Use cc from CFI to load NULL into dataSizeReg for 0 size array
+               iCursor = generateRRFInstruction(cg, TR::InstOpCode::LOCGR, node, dataSizeReg, offsetReg, getMaskForBranchCondition(TR::InstOpCode::COND_BE), true, iCursor);
```
Can you explain why we used getMaskForBranchCondition function call here for storing 0x8 otherwise ?

VermaSh replied:
Just thought it was better for overall readability.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11130,69 +11120,80 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+               iCursor = generateRRInstruction(cg, TR::InstOpCode::XGR, node, offsetReg, offsetReg, iCursor);
+               iCursor = generateRILInstruction(cg, TR::InstOpCode::CFI, node, enumReg, 0, iCursor);
+
+               dataAddrMR = generateS390MemoryReference(resReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg);
```
Just going through the code, do we even need the memory reference variable ? If not may be simply use generateS390MemoryReference in the instruction you are using.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11130,69 +11120,80 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+                  if (comp->getOption(TR_TraceCG))
+                     traceMsg(comp, "Node (%p): Clean out dataAddr field.\n", node);
+
+                  uint16_t bytesToClear = static_cast<uint16_t>(TR::Compiler->om.discontiguousArrayHeaderSizeInBytes() - fej9->getOffsetOfDiscontiguousDataAddrField()) - 1;
```
This is rather confusing. I understand the reason behind subtracting 1, but I think code becomes more readable, if you simply subtract 1 when you are using bytesToClear in the XC instruction generation.

VermaSh commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -4888,13 +4887,18 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
-   // In the mainline, first load the first and second dimensions' lengths into registers.
+   /* In the mainline, first load the first and second dimensions' lengths into registers.
+    *
+    * LGF is used instead of LG so that array size register can be used to NULL dataAddr
```
Ah yes, you are right, I'll update the code.

VermaSh commented:
@r30shah I have updated the PR based on your suggestions, can I please get another review.

VermaSh commented:
> Also can you share a instruction selection log to ensure the ICF is correct ?

I'll share the instruction selection log shortly.

VermaSh commented:
Is this what you were looking for?
**For variable length non 0 size array**:
```
     0x3fcc3df7452 00000062 [     0x3fcbb162460]                                                   Label L0036: # (Start of internal control flow)
     0x3fcc3df7452 00000062 [     0x3fcbb12cfe0] a7 b4 00 eb                                       BRC     BRNP(0xb), Outlined Label L0032, labelTargetAddr=0x000003FCC3DF7628
     0x3fcc3df7456 00000066 [     0x3fcbb12d0c0] a7 0b 00 17                                       AGHI    GPR0,0x17
     0x3fcc3df745a 0000006a [     0x3fcbb12d1a0] a5 07 ff f8                                       NILL    GPR0,0xfff8
     0x3fcc3df745e 0000006e [     0x3fcbb12d340] e3 00 d0 60 00 08                                 AG      GPR0,#415 96(GPR13)
     0x3fcc3df7464 00000074 [     0x3fcbb12d4d0] e3 00 d0 68 00 21                                 CLG     GPR0,#416 104(GPR13)
     0x3fcc3df746a 0000007a [     0x3fcbb12d5a0] a7 24 00 df                                       BRC     BL(0x2), Outlined Label L0032, labelTargetAddr=0x000003FCC3DF7628
     0x3fcc3df746e 0000007e [     0x3fcbb12d740] e3 90 d0 60 00 04                                 LG      GPR9,#417 96(GPR13)
     0x3fcc3df7474 00000084 [     0x3fcbb12d8d0] e3 00 d0 60 00 24                                 STG     GPR0,#418 96(GPR13)
     0x3fcc3df747a 0000008a [     0x3fcbb161960] c0 88 01 bc 3a 00                                 IIHF    GPR8,29112832
     0x3fcc3df7480 00000090 [     0x3fcbb161b30] e3 80 90 00 00 24                                 STG     GPR8,#419 0(GPR9)
     0x3fcc3df7486 00000096 [     0x3fcbb161c70] b9 82 00 11                                       XGR     GPR1,GPR1
     0x3fcc3df748a 0000009a [     0x3fcbb161d50] c2 8d 00 00 00 00                                 CFI     GPR8,0
     0x3fcc3df7490 000000a0 [     0x3fcbb161fe0] 41 00 90 10                                       LA      GPR0,#420 16(GPR9)
     0x3fcc3df7494 000000a4 [     0x3fcbb1620b0] b9 e2 80 01                                       LOCGR   GPR0,0000000000000008,GPR1
     0x3fcc3df7498 000000a8 [     0x3fcbb162190] e3 00 90 08 00 24                                 STG     GPR0,#421 8(GPR9)
     0x3fcc3df749e 000000ae [     0x3fcbb162320] e3 20 91 00 00 36                                 PFD     2, #422 256(GPR9) # Prefetch for store
     0x3fcc3df74a4 000000b4 [     0x3fcbb162b00]                                                   assocreg
     0x3fcc3df74a4 000000b4 [     0x3fcbb162540]                                                   Label L0033: # (End of internal control flow)
```
[method.log](https://github.com/user-attachments/files/18459765/method.log)

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -4948,33 +4948,25 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
-      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, populateFirstDimDataAddrSlot);
-      }
-   else
+   bool isOffHeapAllocationEnabled = TR::Compiler->om.isOffHeapAllocationEnabled();
```
Need proper indentation here.

VermaSh commented:
@r30shah I have updated the PR based on your suggestions. Can I please get another review?

VermaSh commented:
@r30shah Since we only need to clear dataAddr field for off-heap I have moved padding related changes to a separate [PR](https://github.com/eclipse-openj9/openj9/pull/20998). I need to do more testing to verify that my changes don't break 32 bit builds. I'll also look into adding an API to determine whether the padding exists or not.

Can I please get a review for remaining changes?

VermaSh commented:
@r30shah Verified with my changes, we should be good for both clearing padding for 64 bit and 32 bit modes. BTW, I added an assert for `getObjectAlignmentInBytes` in the helper we use to clear array header. Let me know if it clearly states the assumption.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5026,6 +5019,33 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+            node,
+            temp3Reg,
+            generateS390MemoryReference(targetReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg));
+         cursor = generateRXInstruction(cg,
```
Should we use query to get store opcode which returns appropriate opcode for platform (I do understand we may not have off heap for 31 bit, but should use query to use right opcde)

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -5056,26 +5076,27 @@ static TR::Register * generateMultianewArrayWithInlineAllocators(TR::Node *node,
+         // Since 2nd dimension array is 0 length, we can use secondDimLenReg
+         // to clear dataAddr field. secondDimLenReg is expected to be NULL
+         // at this point.
+         cursor = generateRXInstruction(cg, TR::InstOpCode::STG, node, secondDimLenReg, generateS390MemoryReference(temp2Reg, fej9->getOffsetOfDiscontiguousDataAddrField(), cg));
```
Same question as before, should we use query ?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10762,10 +10750,53 @@ genInitArrayHeader(TR::Node * node, TR::Instruction *& iCursor, bool isVariableL
+
+      // Clear 4 byte padding after the size field in the array header.
+      TR_ASSERT_FATAL_WITH_NODE(node,
+         TR::Compiler->om.getObjectAlignmentInBytes() == 8,
```
This is rather long message to print for assert and I am not sure if it makes sense to print that out when someone hits the assert. It does not provide any more value to the failure than having a simple comment here that developer can take a look and see. Also what happens to this assert where using 31 bit ?

VermaSh replied:
Object alignment is 8 bytes on 31 bit as well so this assert will be valid there as well ([minimum object alignment](https://github.com/eclipse-omr/omr/blob/2346911fe109ca9b992896528b28bfdf6a04eb3e/gc/base/ObjectModelBase.hpp#L50-L52)). I think it will be better to check `OMR_MINIMUM_OBJECT_SIZE` rather than object alignment. Looking to see if we have an API for it or if I need to add one. This will help clean up the assert message as well.

r30shah replied:
May be I should ask you this differently. Think about the what is it that you want to prevent with having this Fatal Assert here ?

VermaSh replied:
This was my train of thought behind adding the assert:

We want to avoid corrupting an other object while clearing padding for contiguous and discontiguous header layout. The padding field exists for contiguous header layout in full refs and for discontiguous header layout in compressed refs. But we don't know if the array we are dealing with has the field, we won't know until we do a runtime check. To avoid the extra instruction, we assume that the array has the field and clear memory at corresponding padding offset.

This is where the minimum object size comes into play. For compressed refs, contiguous header is 8 bytes and discontiguous header is 16 bytes, with the padding field at +12 offset. Since we unconditionally clear memory at +12 offset, we could potentially corrupt the object itself (if the array is actually contiguous) or adjacent objects.

We don't need to worry about corrupting the object itself because we are in allocation phase so there is nothing to corrupt at that memory offset. And as long as `OMR_MINIMUM_OBJECT_SIZE` is 16 bytes, we don't need to worry about corrupting any adjacent objects either.

Does that sound good? Should I change the assert and/or add additional details?

VermaSh replied:
Here's the chart I am using for different array header layouts
<img width="975" alt="image" src="https://github.com/user-attachments/assets/9b2cfb33-27cf-4052-8dbc-a8adbcb6b62c" />

r30shah replied:
Thanks a lot for the explanation. What I was asking, is that, we do not need to write this explanation in the ASSERT message when for some reason this condition gets violated. I agree with your reasoning for having the Fatal Assert (To catch the case if things get changed, we catch the issue earlier in testing to prevent accidentally clearing other object). 

I will let you figure out what query would be the correct one here , but recommend you to have a concise trace message in the trace so JIT developer/ service team can go through the code and understand why that assert was thrown.

VermaSh replied:
Ah yes, I agree, I'll push updated assert message shortly.

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11126,73 +11157,93 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+               iCursor = generateRRFInstruction(cg, TR::InstOpCode::LOCGR, node, dataSizeReg, offsetReg, getMaskForBranchCondition(TR::InstOpCode::COND_BE), true, iCursor);
+
+               // Write element address to dataAddr field
+               iCursor = generateRXInstruction(cg,
```
Same question as before, should we use query to get store opcode ?

r30shah commented on runtime/compiler/z/codegen/J9TreeEvaluator.cpp:
```diff
@@ -11126,73 +11157,93 @@ J9::Z::TreeEvaluator::VMnewEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+                  generateS390MemoryReference(resReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg),
+                  iCursor);
+               // Write first data element address to dataAddr field
+               iCursor = generateRXInstruction(cg,
```
same question as before.

VermaSh commented:
@r30shah I have addressed all, expect the object alignment suggestion. I need some more time to fix that.

VermaSh commented:
@r30shah I have updated the assert. I have added a comment explaining the reasoning behind the assert. I'll need additional time to get new APIs to get minimum object allocation size but also don't want to block this PR any longer so was wondering if can get the PR merged and I'll open a separate PR with new API.

VermaSh commented:
@r30shah I found `J9_GC_MINIMUM_OBJECT_SIZE` that we can use to assert minimum object size in bytes. Can I please get another review?

r30shah reviewed (approved):
LGTM

r30shah commented:
jenkins test sanity zlinux jdk11,jdk21

r30shah commented:
Merging  as all tests have passed.
